src/authentication.rs

```diff
@@ -0,0 +1,294 @@
+use async_trait::async_trait;
+
+use crate::error::AuthenticationError;
+
+#[async_trait]
+pub trait Authentication: Send + Sync + 'static {
+    fn auth_method_name(&self) -> String;
+
+    async fn initialize(&mut self) -> Result<(), AuthenticationError>;
+
+    async fn auth_data(&mut self) -> Result<Vec<u8>, AuthenticationError>;
+}
+
+pub mod token {
+    use std::rc::Rc;
+
+    use async_trait::async_trait;
+
+    use crate::authentication::Authentication;
+    use crate::error::AuthenticationError;
+
+    pub struct TokenAuthentication {
+        token: Vec<u8>,
+    }
+
+    impl TokenAuthentication {
+        pub fn new(token: String) -> Rc<dyn Authentication> {
+            Rc::new(TokenAuthentication {
+                token: token.into_bytes()
+            })
+        }
+    }
+
+    #[async_trait]
+    impl Authentication for TokenAuthentication {
+        fn auth_method_name(&self) -> String {
+            String::from("token")
+        }
+
+        async fn initialize(&mut self) -> Result<(), AuthenticationError> {
+            Ok(())
+        }
+
+        async fn auth_data(&mut self) -> Result<Vec<u8>, AuthenticationError> {
+            Ok(self.token.clone())
+        }
+    }
+}
+
+#[cfg(feature = "auth-oauth2")]
+pub mod oauth2 {
+    use std::fmt::{Display, Formatter};
+    use std::fs;
+    use std::time::Instant;
+
+    use async_trait::async_trait;
+    use data_url::{DataUrl};
+    use nom::lib::std::ops::Add;
+    use oauth2::{AuthUrl, ClientId, ClientSecret, Scope, TokenResponse, TokenUrl};
+    use oauth2::AuthType::RequestBody;
+    use oauth2::basic::{BasicClient, BasicTokenResponse};
+    use oauth2::reqwest::async_http_client;
+    use openidconnect::core::CoreProviderMetadata;
+    use openidconnect::IssuerUrl;
+    use serde::Deserialize;
+    use url::Url;
+
+    use crate::authentication::Authentication;
+    use crate::error::AuthenticationError;
+
+    #[derive(Deserialize, Debug)]
+    struct OAuth2PrivateParams {
+        client_id: String,
+        client_secret: String,
+        client_email: Option<String>,
+        issuer_url: Option<String>,
+    }
+
+    #[derive(Deserialize, Debug)]
+    pub struct OAuth2Params {
+        pub issuer_url: String,
+        pub credentials_url: String,
+        pub audience: String,
```

Comment:
Please treat `audience` as optional, since it is an Auth0-specific extension.   Azure, in contrast, uses a scope for the same purpose.